Add "dune describe package-entries"

moyodiallo opened the pull request:
The need is about knowing if a private executable is going to be installed before building the project.

This PR is related to https://github.com/ocurrent/opam-dune-lint/pull/46:
The command `dune describe external-lib` does not give the package name of private executable if it going to be installed, `dune describe external-lib-deps` is used by `opam-dune-lint` to extract the external dependencies.

rgrinberg commented:
The implementation looks fine but how about a clearer name like `package-entries`? `entries` alone doesn't make it clear what is being printed.

moyodiallo commented:
> The implementation looks fine but how about a clearer name like `package-entries`? `entries` alone doesn't make it clear what is being printed.

I buy `package-entries`.

emillon commented:
The implementation is fine, but just to check, does `opam-dune-lint` work with this PR or is this in anticipation? I'd prefer if we have a clear idea of what's necessary in dune rather than add this "just in case". (and if that's all that is necessary, that's perfect)

moyodiallo commented:
> The implementation is fine, but just to check, does `opam-dune-lint` work with this PR or is this in anticipation? I'd prefer if we have a clear idea of what's necessary in dune rather than add this "just in case". (and if that's all that is necessary, that's perfect)

I'm working on that. I had an intention to notify it, after matching the needs of `opam-dune-lint`.

emillon commented:
ok then I'd hold off onto merging these PRs until they've been tested with opam-dune-lint. let me know what that's the case.

moyodiallo commented on src/dune_rules/dune_rules.ml:
```diff
@@ -54,6 +54,10 @@ module Lib_name = Lib_name
 module Diff = Dune_lang.Action.Diff
 module Clflags = Clflags
 
+module Entries = struct
```
You right it could be simple.

moyodiallo commented:
Ready for another review and to be merged.

emillon commented:
has this been tested with opam-dune-lint?

moyodiallo commented on src/dune_rules/install_rules.ml:
```diff
@@ -1140,6 +1140,8 @@ let symlink_rules sctx ~dir =
     | None -> Rules.empty
     | Some rules -> Rules.of_dir_rules ~dir rules )
 
+let stanzas_to_entries = Stanzas_to_entries.stanzas_to_entries
```
OK, good :+1:

emillon commented on bin/describe.ml:
```diff
@@ -704,6 +704,18 @@ module External_lib_deps = struct
     external_resolved_libs setup super_context >>| to_dyn context_name
 end
 
+module Pacakge_entries = struct
```
typo here

moyodiallo commented on src/dune_rules/install.ml:
```diff
@@ -275,6 +288,14 @@ module Entry = struct
+      let open Dyn in
+      let source_to_dyn = function
+        | Dune -> String "dune"
+        | User _ -> String "user"
```
To replace `dune` and `user` you mean ?

rgrinberg commented on test/blackbox-tests/test-cases/describe-package-entries.t:
```diff
@@ -0,0 +1,116 @@
+         (In_build_dir default/foo.ml))
+        (kind file)
+        (dst foo.ml)
+        (section LIB))))
```
I'm curious, why is the section capitalized?

moyodiallo replied:
It is not capitalized by this PR. The section is printed as it is.

emillon replied:
I did some archeology, and it's been added in uppercase in c18541afdee2e8c102552d87267141fa486ce0b6

rgrinberg commented on src/dune_rules/install.ml:
```diff
@@ -258,6 +258,19 @@ module Entry = struct
+    in
+    record
+      [ ("src", Path.Build.to_dyn src)
+      ; ("kind", kind_to_dyn kind)
```
`dyn_of_kind` is our convention

Alizter commented:
FTR I am in the middle of a big refactoring of dune describe. It might make sense to wait for that. Notably you will have to introduce a new subcommand for `package-entries`.

- #7919

emillon commented:
> FTR I am in the middle of a big refactoring of dune describe. It might make sense to wait for that. Notably you will have to introduce a new subcommand for `package-entries`.
> 
> * [split describe into cmdliner subcommands #7919](https://github.com/ocaml/dune/pull/7919)

I wouldn't put an order between those. Both branches exist and a rebase will have to be done, it can be done in a direction or the other.

Alizter commented:
It looks like my branch got merged first. Could you add a `package_entries.ml` file to `bin/describe/` and follow the other describe commands?

moyodiallo committed:
Add "dune describe entries"

    The need is about knowing if a private executable is going to be
    installed before building the project.

Signed-off-by: Alpha DIALLO <moyodiallo@gmail.com>

moyodiallo commented:
I did an upgrade to match the refactoring of `describe`. Would you like give a look ?

@Alizter

Alizter commented on bin/describe/package_entries.ml:
```diff
@@ -0,0 +1,28 @@
+open Import
+open Stdune
+
+let to_dyn =
+  Package.Name.Map.to_dyn (fun entries ->
+      Dyn.List (List.map ~f:Install.Entry.Sourced.to_dyn entries))
```
You can probably just inline this since it is used in one place.

Alizter commented on bin/describe/package_entries.ml:
```diff
@@ -0,0 +1,28 @@
+  let+ common = Common.term
+  and+ context_name = Common.context_arg ~doc:"Build context to use."
+  and+ format = Describe_format.arg
+  and+ _ = Describe_lang_compat.arg in
```
There is no need for a lang compat arg here as this command is new.

Alizter reviewed (approved):
LGTM, can you add a package_entries.mli?

Alizter commented on bin/describe/package_entries.mli:
```diff
@@ -0,0 +1,4 @@
+open Import
+
+(** Dune command to pirnt out information about the entries per package*)
```
```suggestion
(** Dune command to print out information about the entries per package. *)
```

emillon commented on src/install/entry.ml:
```diff
@@ -112,6 +126,14 @@ module Sourced = struct
+  let to_dyn { source; entry } =
+    let open Dyn in
+    let source_to_dyn = function
+      | Dune -> String "dune"
+      | User loc -> Variant ("user", [ Loc.to_dyn loc ])
```
```suggestion
      | Dune -> Variant ("Dune", [])
      | User loc -> Variant ("User", [ Loc.to_dyn loc ])
```

emillon replied:
(you'll need to adapt the consumer on the opam-dune-lint side)

emillon reviewed (commented):
Looks good once the to_dyn is changed!

Commit:
Update src/install/entry.ml

Co-authored-by: Etienne Millon <etienne.millon@gmail.com>
Signed-off-by: Alpha Issiaga DIALLO <alpha@tarides.com>

emillon commented:
Can you promote the tests and add a changelog entry?

moyodiallo committed:
Fix tests

Signed-off-by: Alpha DIALLO <moyodiallo@gmail.com>

Alizter commented on test/blackbox-tests/test-cases/describe-package-entries.t:
```diff
@@ -0,0 +1,218 @@
+Test for the `dune describe package-entries` command
+
+  $ cat >dune-project <<EOF
+  > (lang dune 2.7)
```
Might as well use 3.10.

moyodiallo replied:
you're right.

Alizter commented on bin/describe/package_entries.ml:
```diff
@@ -0,0 +1,24 @@
+open Import
+open Stdune
```
This is already included in Import

moyodiallo replied:
But almost all the files in `bin/describe` include both.

Alizter replied:
Yes, it was recently added to Import I should go and clean those up.

emillon commented:
ping?

moyodiallo committed:
Update changelog

Signed-off-by: Alpha DIALLO <moyodiallo@gmail.com>
